poc_iot_verifier/src/runner.rs

```diff
@@ -154,129 +155,155 @@ impl Runner {
-                        // be discarded from the list returned for the beacon
-                        // thus one of more failing witnesses will not block the overall POC
-                        if !verified_witnesses_result.failed_witnesses.is_empty() {
-                            for failed_witness_report in verified_witnesses_result.failed_witnesses
-                            {
-                                // something went wrong whilst verifying witnesses
-                                // halt here and allow things to be reprocessed next tick
-                                let failed_witness = failed_witness_report.report;
-                                // have to construct the id manually here as dont have the ingest report handy
-                                let id = failed_witness
-                                    .report_id(failed_witness_report.received_timestamp);
-                                Report::update_attempts(&self.pool, &id, Utc::now()).await?;
-                            }
-                            continue;
-                        };
+        // create the struct defining this POC
+        let mut poc =
+            Poc::new(beacon_report.clone(), witnesses.clone(), entropy_start_time).await?;
+
+        // let density_queries = match &self.density_queries {
+        //     Some(density_queries) => density_queries.clone(),
+        //     None => return Err(Error::custom("missing density scaler query sender")),
+        // };
 
-                        let valid_beacon_report = LoraValidBeaconReport {
-                            received_timestamp: beacon_received_ts,
-                            location: beacon_info.location,
-                            hex_scale: beacon_verify_result.hex_scale.unwrap_or_default(),
-                            report: beacon.clone(),
-                        };
-                        self.handle_valid_poc(
-                            beacon,
-                            valid_beacon_report,
-                            verified_witnesses_result,
-                            &lora_valid_poc_tx,
-                            &lora_invalid_witness_tx,
-                        )
+        // verify POC beacon
+        let beacon_verify_result = poc
+            .verify_beacon(hex_density_map.clone(), gateway_cache, &self.pool)
+            .await?;
+        match beacon_verify_result.result {
+            VerificationStatus::Valid => {
+                tracing::debug!(
+                    "valid beacon. entropy: {:?}, addr: {:?}",
+                    beacon.data,
+                    beacon.pub_key.to_string()
+                );
+                // beacon is valid, verify the POC witnesses
+                if let Some(beacon_info) = beacon_verify_result.gateway_info {
+                    let verified_witnesses_result = poc
+                        .verify_witnesses(&beacon_info, hex_density_map, gateway_cache)
```

Comment:
if hex_density_map could be a reference this cloning could be avoided. its especially bad here as we clone for every witness we are verifying